### ISSUE TITLE
Reminder: Support for ARM on OSX

### ISSUE THREAD

**Issue by davidgiga1993** [libgdx/libgdx#6084]
As expected apple just announced that some new mac will have arm cpus instead of intel.
Therefore the native dependency most likely also need to be compiled for apples Axx chips.

Currently there is no publicly available compiler nor documentation (only after paying 500$, thanks apple, see [here](https://www.apple.com/newsroom/2020/06/apple-announces-mac-transition-to-apple-silicon/)). Therefore this task is a reminder to add the new arch as soon as the tools get publicly available.

#### Please select the affected platforms
- [X ] MacOS

**Comment by mustii82** [comment]
We have to wait if the JVM comes to the new Mac ARM Architetcure, Im sure it will.

**Comment by crykn** [comment]
Just to provide some resources on this: JDK support for Windows ARM is done ([JEP-338](https://openjdk.java.net/jeps/388)) and support for MacOS ARM underway ([JEP-391](http://openjdk.java.net/jeps/391)). Also important for us is the LWJGL 3 support, which is apparently progressing with the help of Microsoft, see [here](https://github.com/LWJGL/lwjgl3/issues/601).

**Comment by MrStahlfelge** [comment]
Given the hassles that MacOS support is, I am very curious if it is better to use RoboVM to support ARM Macs. Did anyone try?

**Comment by PokeMMO** [comment]
Zulu has released a macOS ARM JDK.
https://www.azul.com/downloads/zulu-community/?os=macos&architecture=arm-64-bit&package=jdk

**Comment by crykn** [comment]
Apparently the latest snapshot of LWJGL now support the macOS ARM architecture. More info is available in the issue linked in my comment above.

**Comment by kishtner** [comment]
Are there any updates on this? I would like to use LWJGL3 on my M1 Mac mini but I get the following error: "Cocoa: Failed to find service port for display"

Is there a workaround for the time being?

**Comment by crykn** [comment]
Hey @kishtner, we're still waiting for LWJGL 3.3.0 to be released. On our end we're all set since  https://github.com/libgdx/libgdx/commit/7ac239c81f7d126c3be5ad18c25bcdb2ff811fd5.

**Comment by fgnm** [comment]
If someone want to try ARM support on macOS without waiting LWJGL `3.3.0` to be released can modify gradle dependencies... This is my setup:
```Groovy
dependencies {
    //Include LWJGL3 Backend
    implementation "com.badlogicgames.gdx:gdx-backend-lwjgl3:$gdxVersion"
    //Override LWJGL3 version
    implementation "org.lwjgl:lwjgl:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl:$lwjgl3Version:natives-windows-x86"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl-glfw:$lwjgl3Version:natives-windows-x86"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl-jemalloc:$lwjgl3Version:natives-windows-x86"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl-opengl:$lwjgl3Version:natives-windows-x86"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl-openal:$lwjgl3Version:natives-windows-x86"
    
    //Additional LWJGL3 modules
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-linux"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-linux-arm32"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-linux-arm64"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-macos"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-macos-arm64"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-windows"
    implementation "org.lwjgl:lwjgl-tinyfd:$lwjgl3Version:natives-windows-x86"
}
```
Version used
```
gdxVersion = 1.10.1-SNAPSHOT
lwjgl3Version = 3.3.0-SNAPSHOT
```

**Comment by metaphore** [comment]
[LWJGL 3.3.0](https://github.com/LWJGL/lwjgl3/releases/tag/3.3.0) has just been released. It comes along with arm64 support for macOS and Windows 👌

**Comment by tommyettinger** [comment]
I just tried 1.10.1-SNAPSHOT of libGDX with 3.3.0 of LWJGL, and some things seem to still not quite work on an M1. I don't have the most detailed specifics of the hardware and JDK installation, but I believe the user with an M1 has an ARM64 JDK. The issue is with libgdx64.dylib ; SharedLibraryLoader seems to be trying to use an x86_64 dylib when it needs an arm64e dylib. I could easily be doing something wrong in the dependencies; my current build.gradle file is https://github.com/tommyettinger/gdx-liftoff/blob/master/build.gradle#L61-L110 (properties are defined in gradle.properties, next to that file). The error the user reported is:
```
Exception in thread "main" com.badlogic.gdx.utils.GdxRuntimeException: Couldn't load shared library 'libgdx64.dylib' for target: Mac OS X, 64-bit
    at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:120)
    at com.badlogic.gdx.utils.GdxNativesLoader.load(GdxNativesLoader.java:31)
    at com.badlogic.gdx.backends.lwjgl3.Lwjgl3NativesLoader.load(Lwjgl3NativesLoader.java:28)
    at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.initializeGlfw(Lwjgl3Application.java:80)
    at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:91)
    at gdx.liftoff.MainKt.main(main.kt:104)
    at gdx.liftoff.MainKt.main(main.kt)
Caused by: com.badlogic.gdx.utils.GdxRuntimeException: java.lang.UnsatisfiedLinkError: /private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib: dlopen(/private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib, 0x0001): tried: '/private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libgdx64.dylib' (no such file)
    at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:316)
    at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:116)
    ... 6 more
Caused by: java.lang.UnsatisfiedLinkError: /private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib: dlopen(/private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib, 0x0001): tried: '/private/var/folders/g8/hsy5g3915r51l4jr4jgvb_h00000gn/T/libgdxhacker/f4ccdf55/libgdx64.dylib' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/libgdx64.dylib' (no such file)
    at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
    at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:384)
    at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:228)
    at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:170)
    at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2389)
    at java.base/java.lang.Runtime.load0(Runtime.java:755)
    at java.base/java.lang.System.load(System.java:1953)
    at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:322)
    at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:291)
    ... 7 more
```
On one of the first lines, it says the target is `Mac OS X, 64-bit` -- this seems wrong now that there are x86_64 and arm64e that are both 64-bit but incompatible with each other.

**Comment by PokeMMO** [comment]
That's interesting, it should be trying to load `libgdxarm64.dylib`

**Comment by tommyettinger** [comment]
The gdx-jnigen-loader version is 2.2.0 .

I don't have access to the M1 Mac in question, but I have the JAR that had the issue -- I can upload it here, I think.
[gdx-liftoff-1.10.0.9-SNAPSHOT.jar.zip](https://github.com/libgdx/libgdx/files/7573897/gdx-liftoff-1.10.0.9-SNAPSHOT.jar.zip)

Hope this helps.

**Comment by PokeMMO** [comment]
You somehow have the wrong version of `SharedLibraryLoader` in that jar.

It should look like this https://github.com/libgdx/gdx-jnigen/blob/b807cce413585da1decc494dc0bb56e7bbc0694e/gdx-jnigen-loader/src/main/java/com/badlogic/gdx/utils/SharedLibraryLoader.java#L107

But the decompiled code is still only 
```
return "lib" + libraryName + (SharedLibraryLoader.is64Bit ? "64.dylib" : ".dylib");
```

**Comment by tommyettinger** [comment]
That's certainly strange, because when I open up the JAR in IDEA, it has this line in that method (still 107):
```java
if (isMac) return "lib" + libraryName + (isARM ? "arm" : "") + (is64Bit ? "64.dylib" : ".dylib");
```
But when decompiled, it's on line 70, and has the contents
```java
return isMac ? "lib" + libraryName + (is64Bit ? "64.dylib" : ".dylib") : libraryName;
```

This suggests there's a mismatch between the JAR artifact uploaded for version 2.2.0 and the sources, or somehow my dependencies are out of sorts and an older jnigen is being used.

**Comment by tommyettinger** [comment]
OK, this is strange. It appears as if the libGDX 1.10.1-SNAPSHOT jar includes an old version of SharedLibraryLoader, and this overwrites gdx-jnigen-loader's version. I don't know if this is just my snapshot jar getting mixed up in some way, but I'll download the current version from Sonatype and see what it has.

**Comment by mayt-moonboy** [comment]
I am also having this issue. Is there an official fix for this? My build is with bazel so it is strange that I am having this issue with source that previously worked.

**Comment by tommyettinger** [comment]
@mayt-moonboy Using the highlighted dependencies here https://github.com/tommyettinger/gdx-liftoff/blob/master/build.gradle#L60-L100 worked once lwjgl3Version was defined to 3.3.0 and gdxVersion to 1.10.1-SNAPSHOT . I needed to remove the mavenLocal() repo in my case, but only because I had a much older self-installed libGDX version -- this probably won't be needed for you. If it is, you can also delete libGDX at version 1.10.1-SNAPSHOT from your ~/.m2/ repo. I don't use Bazel, nor does anyone else active in the libGDX community that I know of, but there's probably a way to make your Bazel dependencies act equivalently to the linked Gradle dependencies. The officially supported way uses Gradle, so... good luck, I hope it isn't too hard to translate the Gradle code to Bazel code.

**Comment by mayt-moonboy** [comment]
@tommyettinger thanks for the reply. Unfortunately this just resulted in another issue. Some important things to note is that my build had worked perfectly previously. I had a local cache of my dependencies until I recently checked out a fresh version of the project and now it fails with the error listed above (as well as this new one after following your advice I will list below). 

I am curious why the build would be failing all of the sudden. Did some upstream dependency get updated? Do release versions of libgdx typically leverage snapshot dependencies? This definitely feels like an issue where some dependency version was not locked in.

```log
java.lang.NoSuchMethodError: org.lwjgl.openal.AL.create()V
    at com.badlogic.gdx.backends.lwjgl.audio.OpenALLwjglAudio.<init>(OpenALLwjglAudio.java:72)
    at com.badlogic.gdx.backends.lwjgl.LwjglApplication.createAudio(LwjglApplication.java:282)
    at com.badlogic.gdx.backends.lwjgl.LwjglApplication.<init>(LwjglApplication.java:90)
    at com.badlogic.gdx.backends.lwjgl.LwjglApplication.<init>(LwjglApplication.java:71)
    at <redacted>game.engine.demo.desktop.MainApplicationImpl.<init>(MainApplicationImpl.kt:12)
    at <redacted>game.engine.demo.desktop.MainApplicationImpl_Factory.newInstance(MainApplicationImpl_Factory.java:35)
    at <redacted>game.engine.demo.desktop.MainApplicationImpl_Factory.get(MainApplicationImpl_Factory.java:25)
    at <redacted>game.engine.demo.desktop.MainApplicationImpl_Factory.get(MainApplicationImpl_Factory.java:8)
    at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
    at <redacted>game.engine.demo.desktop.DaggerMainComponent.getMainApplication(DaggerMainComponent.java:358)
    at <redacted>game.engine.demo.desktop.MainComponent$Companion.getMainApplication(MainComponent.kt)
    at <redacted>game.engine.demo.desktop.MainKt.main(main.kt:6)
    Exception in thread "LWJGL Application" java.lang.UnsatisfiedLinkError: /private/var/folders/5x/c78kkzp50vb127c44dhbxjyc0000gn/T/libgdx<redacted>/f7649394/liblwjgl.dylib: dlopen(/private/var/folders/5x/c78kkzp50vb127c44dhbxjyc0000gn/T/libgdx<redacted>/f7649394/liblwjgl.dylib, 0x0001): tried: '/private/var/folders/5x/c78kkzp50vb127c44dhbxjyc0000gn/T/libgdx<redacted>/f7649394/liblwjgl.dylib' (fat file, but missing compatible architecture (have 'x86_64', need 'arm64e')), '/usr/lib/liblwjgl.dylib' (no such file)
at java.lang.ClassLoader$NativeLibrary.load(Native Method)
at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1950)
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1832)
at java.lang.Runtime.load0(Runtime.java:811)
at java.lang.System.load(System.java:1088)
at org.lwjgl.Sys$1.run(Sys.java:70)
at java.security.AccessController.doPrivileged(Native Method)
at org.lwjgl.Sys.doLoadLibrary(Sys.java:66)
at org.lwjgl.Sys.loadLibrary(Sys.java:96)
at org.lwjgl.Sys.<clinit>(Sys.java:117)
at org.lwjgl.opengl.Display.<clinit>(Display.java:135)
at com.badlogic.gdx.backends.lwjgl.LwjglGraphics.setVSync(LwjglGraphics.java:643)
at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:125)
```

**Comment by tommyettinger** [comment]
@mayt-moonboy You can't mix LWJGL3 and LWJGL2, and you're using LWJGL 2. There have been no updates to LWJGL2 in nearly 7 years, and of course ARM Macs are newer than that. If you want native support for M1 Macs, you need to change to LWJGL3. https://libgdx.com/news/2021/07/devlog-7-lwjgl3 covers the specifics; if you have more questions about the process, feel free to ask on the Discord, since an interactive exchange like that makes more sense than here. https://libgdx.com/community/discord/

**Comment by mayt-moonboy** [comment]
@tommyettinger It looks like that is possibly coming from libgdx-tools which has a dependency on lwjgl2. I am going to try removing my dependency on tools.

**Comment by mayt-moonboy** [comment]
Ok, that worked. I do have to now finegle some of my settings which were geared towards an LwjglApplicationConfiguration. Thanks for the help!
